deckhouse-controller/pkg/controller/module-controllers/release/controller.go

```diff
@@ -526,21 +521,33 @@ func (c *moduleReleaseReconciler) reconcilePendingRelease(ctx context.Context, m
 	}
 
 	err = releaseUpdater.ApplyPredictedRelease(windows)
-	if errors.Is(err, updater.ErrNotReadyForDeploy) {
-		//TODO: create custom error type with additional fields like reason end requeueAfter
-		return ctrl.Result{RequeueAfter: defaultCheckInterval}, nil
-	}
 	if err != nil {
-		return ctrl.Result{RequeueAfter: defaultCheckInterval}, fmt.Errorf("apply predicted release: %w", err)
+		return r.wrapApplyReleaseError(err)
 	}
 
 	modulesChangedReason = "a new module release found"
-	return c.cleanUpModuleReleases(ctx, mr)
+	return r.cleanUpModuleReleases(ctx, mr)
+}
+
+func (r *moduleReleaseReconciler) wrapApplyReleaseError(err error) (ctrl.Result, error) {
+	var notReadyErr *updater.NotReadyForDeployError
+	if errors.As(err, &notReadyErr) {
+		r.logger.Infoln(err.Error())
+		// TODO: requeue all releases if deckhouse update settings is changed
+		// requeueAfter := notReadyErr.RetryDelay()
+		// if requeueAfter == 0 {
+		// requeueAfter = defaultCheckInterval
+		// }
+		// r.logger.Infof("%s: retry after %s", err.Error(), requeueAfter)
+		return ctrl.Result{Requeue: true, RequeueAfter: notReadyErr.RetryDelay()}, nil
+	}
+
+	return ctrl.Result{RequeueAfter: defaultCheckInterval}, fmt.Errorf("apply predicted release: %w", err)
```

Comment:
If the returned error is non-nil, the Result is ignored and the request will be requeued using exponential backoff. The only exception is if the error is a TerminalError in which case no requeuing happens.